### ISSUE TITLE
Support color profiles

### ISSUE THREAD

**Issue by LukasKlement** [Kozea/WeasyPrint#844]
Is there a way to change the colour profile to CMYK? Whilst it's easy to use PIL for jpgs, I haven't found a way to do it for entire PDFs.

There will be cmyk() support in CSS4, but that's still some time to go: https://www.w3.org/TR/css-color-4/#cmyk-colors

**Comment by liZe** [comment]
> Is there a way to change the colour profile to CMYK? Whilst it's easy to use PIL for jpgs, I haven't found a way to do it for entire PDFs.

Cairo (the library used in WeasyPrint to generate PDF and PNG files) doesn't support CMYK. Here's a [sad post](http://libregraphicsworld.org/blog/entry/color-managed-cairo-is-slipping-away) about it, it's old but still relevant.

**Comment by LukasKlement** [comment]
Perhaps we could tackle this by converting the colour space after generating a PDF? I've seen an approach on this here: https://github.com/johndoe31415/pdfminify/blob/master/llpdf/filters/PDFAFilter.py

This might help us making some progress on this ticket as well: https://github.com/Kozea/WeasyPrint/issues/630

**Comment by liZe** [comment]
> Perhaps we could tackle this by converting the colour space after generating PDFs?

Yes, we can do that, but only if we use colors that can have a RGB representation in a bijective way (as we'll have to use cairo at one moment, and cairo can only store RGB colors).

> I've seen an approach on this here
> This might help us making some progress on this ticket as well: #630

pdfminify? What an interesting project, wow. You're right, it may be really helpful.

**Comment by liZe** [comment]
Related to #1091.

**Comment by dominikhuetter** [comment]
Is there any update on this issue?

**Comment by liZe** [comment]
> Is there any update on this issue?

Now that we use our own PDF generator, it could be possible to support more features about color profiles. That’s not planned for the next milestones, but it may be in the future.

**Comment by liZe** [comment]
> Could you point to the relevant code sections where CMYK support would fit?

That’s a very good question. I don’t know the PDF specification well enough to know how to generate a PDF with color profiles, so it’s pretty hard to tell where the relevant code is…

**Comment by grewn0uille** [comment]
Hello!

As it’s soon our 2-year anniversary as CourtBouillon, we opened a [short survey](https://thym.courtbouillon.org/1) to know more about your expectations.
Don’t hesitate to support this feature and give it a boost 🚀!

The survey will be opened until October 10th.

**Comment by TellersFella** [comment]
Hey, I just looked at the survey results and color profiles seem relevant, but not at the absolute top of the list. So my question is, is there anything planned regarding this?

**Comment by davidmarogy** [comment]
Hello @liZe,

i just wanted to ask if there is any new updates regarding implementation of ICC-based colors or any CMYK conversion in Weasyprint?
 
Best regards,
David

**Comment by liZe** [comment]
> i just wanted to ask if there is any new updates regarding implementation of ICC-based colors or any CMYK conversion in Weasyprint?

Yes, some work has been done, but we’re not close to a proper CMYK support in WeasyPrint yet.

What we’ve been working on is the support of [CSS color level 4 in tinycss2](https://github.com/Kozea/tinycss2/issues/47). This feature will enable the possibility to define non-RGB colors using CSS. The work is not merged yet (mainly because we’ve got a lot of work on other parts of WeasyPrint), but hopefully most of the code is already written. Tests and ideas to solve the latest missing features are welcome!

The next step is to include this support in WeasyPrint. It means that we’ll have to manipulate abstract colors instead of RGB colors everywhere, and that we’ll have to write the correct PDF data according to the space of the color. Images are another topic we have to take care of. That’s not done at all.

And we didn’t even talk of ICC and CMYK yet!

These features are possible with [CSS color level 5](https://drafts.csswg.org/css-color-5/), thanks to [`@color-profile`](https://drafts.csswg.org/css-color-5/#at-profile). We need to add the support of level 4 in tinycss2, and then in WeasyPrint. It should be easier than adding level 4 support though, because we’ll already have done the work to support different color spaces.

**Comment by grewn0uille** [comment]
Hello!

As you may know, two weeks ago was [CourtBouillon 3-year anniversary](https://www.courtbouillon.org/blog/00046-three-years-of-courtbouillon) 🎉.

For this occasion, we prepared [a short survey](https://thym.courtbouillon.org/4) to have your opinion on this year’s features and to know what you’d like to see in the future!
Don’t hesitate to give a boost to this feature ✨️

 The survey is opened until November 19.

**Comment by liZe** [comment]
Fixed by #2560.
